### PR TITLE
feat(clients): use custom cognito-identity and pinpoint clients

### PR DESCRIPTION
**Pull request by AllanZhengYP**
#### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1632,6 +1632,7 @@ releasable_branches: &releasable_branches
       - release
       - main
       - next
+      - v5/custom-clients
```
Comment (AllanZhengYP) [inline .circleci/config.yml:1635]: This line is still needed in v5/custom-clients branch. Will remove this in a follow-up PR

File: packages/analytics/src/Providers/EventBuffer.ts [hunk 2]
```diff
@@ -167,50 +159,53 @@ export default class EventsBuffer {
 	}
 
 	private _processPutEventsSuccessResponse(
-		data: PutEventsResponse,
+		data: PutEventsOutput,
 		eventMap: EventMap
 	) {
-		const { Results } = data.EventsResponse;
+		const { Results = {} } = data.EventsResponse ?? {};
```
Comment (stocaaro) [inline packages/analytics/src/Providers/EventBuffer.ts:165]: Would this destructuring be clearer if the defaults where consolidated to one side of the assignment? eg `const { Results } = data.EventsResponse ?? {Results: {}};`

File: packages/analytics/src/Providers/EventBuffer.ts [hunk 3]
```diff
@@ -167,50 +159,53 @@ export default class EventsBuffer {
...
+				if (StatusCode && ACCEPTED_CODES.includes(StatusCode)) {
+					eventObject.handlers.resolve(response);
+					return;
+				}
 
-					if (RETRYABLE_CODES.includes(StatusCode)) {
-						retryableEvents.push(eventObject);
-						return;
-					}
+				if (StatusCode && RETRYABLE_CODES.includes(StatusCode)) {
```
Comment (stocaaro) [inline packages/analytics/src/Providers/EventBuffer.ts:197]: Is `includes` unhappy given `undefined`? Seems no-op, but its not clear to me if this is driven by ts complaining or something else.

File: packages/api-graphql/package.json [hunk 4]
```diff
@@ -64,7 +64,7 @@
       "name": "API (GraphQL client)",
       "path": "./lib-esm/index.js",
       "import": "{ Amplify, GraphQLAPI }",
-      "limit": "104.6 kB"
+      "limit": "87.1 kB"
```
Comment (stocaaro) [inline packages/api-graphql/package.json:67]: ❤️

File: packages/core/src/clients/handlers/fetch.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+import 'isomorphic-unfetch'; // TODO: remove this dependency in v6
+import { HttpRequest, HttpResponse, HttpTransferOptions } from '../types/http';
+import { TransferHandler } from '../types/core';
+
+const shouldSendBody = (method: string) =>
+	!['HEAD', 'GET', 'DELETE'].includes(method.toUpperCase());
+
+export const fetchTransferHandler: TransferHandler<
```
Comment (stocaaro) [inline packages/core/src/clients/handlers/fetch.ts:11]: Seems like this is the heart and soul of the custom client.  Seems deceptively simple.

File: packages/core/__tests__/AwsClients/Pinpoint/updateEndpoint-test.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	mockFailureResponse,
+	mockJsonResponse,
+	mockRequestId,
+	pinpointHandlerOptions,
+} from '../testUtils/data';
+
+jest.mock('../../../src/clients/handlers/fetch');
+
+// API reference: https://docs.aws.amazon.com/pinpoint/latest/apireference/apps-application-id-endpoints-endpoint-id.html#UpdateEndpoint
+describe('Pinpoint - updateEndpoint', () => {
```
Comment (stocaaro) [inline packages/core/__tests__/AwsClients/Pinpoint/updateEndpoint-test.ts:23]: The different API tests are a bit repetitive. Don't mind this in a test if its telling the story, but we probably have more lines than we need to tell the story that a given input results in a given output or error when the API is called depending on conditions.

Comment (stocaaro) [reply packages/core/__tests__/AwsClients/Pinpoint/updateEndpoint-test.ts:23]: Might be worth considering a refactor as we add more apis to keep the tests as succinct and on point as possible.

File: packages/core/__tests__/clients/utils/isClockSkewedError-test.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { isClockSkewError } from '../../../src/clients/utils/isClockSkewError';
+
+describe('isClockSkewError', () => {
```
Comment (stocaaro) [inline packages/core/__tests__/clients/utils/isClockSkewedError-test.ts:6]: Seems like there could be a lot of corners like this. How do you know you've captured/explored all of them?  I'm guessing there's a way you know this from all your work that I'm not thinking of as a reviewer coming to it late in the process.

**Review by stocaaro — COMMENTED** [review]
I'm only half way through, submitting comments now and will return to this in a couple hours.

File: packages/core/__tests__/clients/utils/isClockSkewedError-test.ts (hunk 7 above)
Comment (AllanZhengYP) [reply packages/core/__tests__/clients/utils/isClockSkewedError-test.ts:6]: @stocaaro They are referred from the AWS SDK. The reference is already included in the comment: [diff](https://github.com/aws-amplify/amplify-js/pull/11427/files#diff-a9a79d01515941bf3bceeb0f465446b9ca6ed29811498f732ff0a3a76da3a4a4R4)

File: packages/core/__tests__/AwsClients/Pinpoint/updateEndpoint-test.ts (hunk 6 above)
Comment (AllanZhengYP) [reply packages/core/__tests__/AwsClients/Pinpoint/updateEndpoint-test.ts:23]: We can definitely consider having a test runner looping through the test cases. I explored it a little bit and it's not as easy to maintain is one thought. Maybe it pays off when we add more clients in the future. For now I can add an action item for later.

File: packages/core/src/clients/handlers/fetch.ts (hunk 5 above)
Comment (AllanZhengYP) [reply packages/core/src/clients/handlers/fetch.ts:11]: More native interfaces, Less code, less bundle size 😁

File: packages/analytics/src/Providers/EventBuffer.ts (hunk 3 above)
Comment (AllanZhengYP) [reply packages/analytics/src/Providers/EventBuffer.ts:197]: It is from TS complaining. You are not allow to check `undefined` from `number[]`

File: package.json [hunk 8]
```diff
@@ -28,6 +29,7 @@
 		"publish:next": "lerna publish --canary --force-publish \"*\" --yes --dist-tag=next --preid=next${PREID_HASH_SUFFIX} --exact --no-verify-access",
 		"publish:release": "lerna publish --conventional-commits --yes --message 'chore(release): Publish [ci skip]' --no-verify-access",
 		"publish:verdaccio": "lerna publish --no-push --canary minor --dist-tag=unstable --preid=unstable --exact --force-publish --yes --no-verify-access",
+		"publish:v5/custom-clients": "lerna publish --canary --force-publish \"*\" --yes --dist-tag=v5-custom-clients --preid=v5-custom-clients${PREID_HASH_SUFFIX} --exact --no-verify-access",
```
Comment (AllanZhengYP) [inline package.json:32]: To be removed in follow up PR.

File: packages/core/src/Credentials.ts [hunk 9]
```diff
@@ -481,12 +458,10 @@ export class CredentialsClass {
...
-			} = await cognitoClient.send(
-				new GetCredentialsForIdentityCommand({
-					IdentityId: guestIdentityId || generatedOrRetrievedIdentityId,
-					Logins: logins,
-				})
-			);
+			} = await getCredentialsForIdentity(cognitoConfig, {
+				IdentityId: guestIdentityId || generatedOrRetrievedIdentityId,
+				Logins: logins,
+			});
```
Comment (stocaaro) [inline packages/core/src/Credentials.ts:464]: I think this is the first point I recognized that we're replacing a client command execution with single functions that operate as both. That seems fine, makes me wonder how this will change mocking in tests and similar. Do our customers mock across these client interfaces?

File: packages/core/src/clients/internal/composeServiceApi.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+			...config,
+		} as unknown as TransferHandlerOptions & ServiceClientOptions;
+		const endpoint = await resolvedConfig.endpointResolver({
+			region: resolvedConfig.region,
+		});
+		// Unlike AWS SDK clients, a serializer should NOT populate the `host` or `content-length` headers.
+		// Both of these headers are prohibited per Spec(https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_header_name).
+		// They will be populated automatically by browser, or node-fetch polyfill.
+		const request = serializer(input, endpoint);
+		const response = await transferHandler(request, {
```
Comment (stocaaro) [inline packages/core/src/clients/internal/composeServiceApi.ts:41]: If I'm reading this right, there essentially isn't a client anymore, right?  These function composition API methods essentially take the place of clients.  Seems cool, interested to see how this changes some of the test mocking I've been working with. Probably examples of test mocking updates in this change that I haven't thought through yet.

File: packages/notifications/src/InAppMessaging/Providers/AWSPinpointProvider/utils.ts [hunk 11]
```diff
@@ -2,10 +2,7 @@
 // SPDX-License-Identifier: Apache-2.0
 
 import { Amplify, ConsoleLogger, Hub } from '@aws-amplify/core';
-import {
-	InAppMessageCampaign as PinpointInAppMessage,
-	Layout as PinpointInAppMessageLayout,
```
Comment (stocaaro) [inline packages/notifications/src/InAppMessaging/Providers/AWSPinpointProvider/utils.ts:7]: Looks like we have layouts still in our own code now, why not use it as reference instead of conditioning against constants? Downside is extra import, upside is single integration point where the underlying value can change without breaking the behavior.

File: package.json [hunk 12]
```diff
@@ -18,6 +18,7 @@
 		"build:watch": "concurrently 'lerna run build:cjs:watch --parallel' 'lerna run build:esm:watch --parallel' --raw",
 		"build:esm:watch": "lerna run build:esm:watch --parallel",
 		"build:cjs:watch": "lerna run build:cjs:watch --parallel",
+		"build:client-types": "cd scripts/dts-bundler && yarn && yarn run build",
```
Comment (stocaaro) [inline package.json:21]: Does this rebuild some of the files checked in with this change?  Thats cool, if thats what this is. When as a team will we perform this operation?  If this isn't whats happening here, what is this, have we documented what it is and when we should use it somewhere?

File: packages/core/src/Credentials.ts (hunk 9 above)
Comment (AllanZhengYP) [reply packages/core/src/Credentials.ts:464]: Customers should mock the Amplify library instead of underlying AWS SDK clients, which is our implementation detail. Whether and how we are using AWS SDK is never our public interface.

File: package.json (hunk 12 above)
Comment (AllanZhengYP) [reply package.json:21]: Yes they are generating the `core/src/AwsClients/*/types.ts` like [the one for Pinpoint](https://github.com/aws-amplify/amplify-js/pull/11427/files#diff-d4104f7acd2931d26c4b5f614e65eb5671c24c08cc5baf0a684eaf1d5e21d2f7). And the questions are answered in the [folder's README](https://github.com/aws-amplify/amplify-js/pull/11427/files#diff-ad126c534dfec1e5ce8705e8e144831d0bcdc26fe084612134440b88d2fe4569).

File: packages/notifications/src/InAppMessaging/Providers/AWSPinpointProvider/utils.ts (hunk 11 above)
Comment (AllanZhengYP) [reply packages/notifications/src/InAppMessaging/Providers/AWSPinpointProvider/utils.ts:7]: This worry is unnecessary. The `Layout` is actually modeled [in service API as string constants](https://docs.aws.amazon.com/pinpoint/latest/apireference/apps-application-id-endpoints-endpoint-id-inappmessages.html#:~:text=app%20messaging%20payload.-,Layout,-string). The SDK uses enum for their ergonomic considerations. Regarding changing the underlying constant values, it would be an APIbreaking change on server side anyway and clients should not be used to detect server-side breaking change.

File: packages/core/src/AwsClients/CognitoIdentity/types.ts [hunk 13]
```diff
@@ -0,0 +1,159 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Generated by scripts/dts-bundler/README.md
```
[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply packages/core/src/AwsClients/CognitoIdentity/types.ts:4]: It's not copied but generated from the pinned AWS SDK clients(3.335.0) now. And this is the backlink to the readme on how to generate this file.

File: packages/core/src/AwsClients/CognitoIdentity/getCredentialsForIdentity.ts [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+	HttpRequest,
+	HttpResponse,
+	composeServiceApi,
+	parseJsonBody,
+	parseJsonError,
+	parseMetadata,
+} from '../../clients';
+import type {
+	GetCredentialsForIdentityCommandInput as GetCredentialsForIdentityInput,
+	GetCredentialsForIdentityCommandOutput as GetCredentialsForIdentityOutput,
```
[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply packages/core/src/AwsClients/CognitoIdentity/getCredentialsForIdentity.ts:21]: Because we are not using the command & client interface but a functional interface, removing them reduces the incorrect similarity to the AWS SDK structure.

File: package.json (hunk 12 above)
Comment (stocaaro) [reply package.json:21]: The Readme doesn't speak to when this should be run. Should we regenerate this every release? Maybe just include this in the build process automation and not check the files in? If these change infrequently such that we wouldn't want these slowing down build automation, then under what condition might we want to regenerate them? If its manual, I want to know how we're going to know we need to do it, right?

File: packages/notifications/src/InAppMessaging/Providers/AWSPinpointProvider/utils.ts (hunk 11 above)
Comment (stocaaro) [reply packages/notifications/src/InAppMessaging/Providers/AWSPinpointProvider/utils.ts:7]: Sure. I've been an advocate for using string literals over enums, but our team has elected to stick with enums. Since we're a team that uses enums, why aren't we using the enums here / why are we moving away from using them explicitly?

File: packages/core/src/Credentials.ts (hunk 9 above)
Comment (stocaaro) [reply packages/core/src/Credentials.ts:464]: Should they? Do we document this for them in a way that indicates that we support mocking the library and not mocking the underlying clients?

In a prior life I've been a loud advocate for mocking as close to the network boundary as is reasonably possible, which would be the client or lower in our case, right? How certain are we that customers aren't doing this?

File: packages/core/src/AwsClients/CognitoIdentity/getCredentialsForIdentity.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (stocaaro) [reply packages/core/src/AwsClients/CognitoIdentity/getCredentialsForIdentity.ts:21]: Got it. To restate to confirm my understanding...

We are moving away from the SDK interface where Clients receive Commands through `send` calls with functions that couple together client configuration and command execution into single callable actions. We are using the SDK code generation logic to generate the input values and because it comes from the SDK it uses the Command language in type naming. In this instance we are using `as` to rename these for clarity to avoid confusion with the Client/Command interface.

Is my understanding correct or am I confused in a way that you can help me clarify?

File: package.json (hunk 12 above)
Comment (AllanZhengYP) [reply package.json:21]: Good question. We don't need to regenerate them often. It only needs to be run when we need to update the checked-in types, like updating or adding a new API(i.e headObject). It's a script that eliminate the manual work of copying types from AWS SDK to Amplify JS library. The reason I don't include it to build process because:
1. Generated types needs to be reviewed
2. Maintainers need these types to be checked in to work with library.
3. The generated types will be the same as long as it generated from the same pinned SDK version.

**Review by Jshhhh — DISMISSED** [review]

File: package.json (hunk 12 above)
Comment (stocaaro) [reply package.json:21]: I think 1 is a good reason. Re: 2, you need to build the library to work on it already (eg `setup-dev`) so I don't think this holds. 3 Seems like a problem. We need to get away from using a pinned version of the SDK, right?

File: packages/notifications/src/InAppMessaging/Providers/AWSPinpointProvider/utils.ts (hunk 11 above)
Comment (AllanZhengYP) [reply packages/notifications/src/InAppMessaging/Providers/AWSPinpointProvider/utils.ts:7]: It's because in enum is actual JS code instead of types. Sync'ing them between AWS SDK and Amplify especially in public interface will be a headache. Using string literals does not affect any usability, but easier to maintain. 
As we talked offline, I will add followup action item to spot scheduled breaking change in our CI.

**Review by stocaaro — DISMISSED** [review]
LGTM

File: packages/core/src/Credentials.ts [hunk 15]
```diff
@@ -407,26 +381,33 @@ export class CredentialsClass {
...
 
-		const cognitoClient = createCognitoIdentityClient({
-			region: identityPoolRegion || region,
-		});
+		const cognitoConfig = { region: identityPoolRegion ?? region };
 
-		let credentials = undefined;
-		if (identity_id) {
-			const cognitoIdentityParams: FromCognitoIdentityParameters = {
+		const credentialsProvider = async () => {
```
Comment (elorzafe) [inline packages/core/src/Credentials.ts:386]: nit: the name could be `authenticatedCredentialsProvider`

File: packages/core/src/Credentials.ts [hunk 16]
```diff
@@ -291,47 +280,29 @@ export class CredentialsClass {
...
-
-			Note: Retreive identityId from CredentialsProvider once aws-sdk-js v3 supports this.
-			*/
-			const credentialsProvider: CredentialProvider = async () => {
-				const { IdentityId } = await cognitoClient.send(
-					new GetIdCommand({
-						IdentityPoolId: identityPoolId,
-					})
-				);
+		const credentialsProvider = async () => {
```
Comment (elorzafe) [inline packages/core/src/Credentials.ts:285]: nit: I would name it `guestCredentialsProvider`

semi-nit: This code is very similar to a code below

File: packages/core/src/AwsClients/Pinpoint/types.ts [hunk 17]
```diff
@@ -0,0 +1,844 @@
+/* tslint:disable:max-line-length */
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Generated by scripts/dts-bundler/README.md
```
Comment (elorzafe) [inline packages/core/src/AwsClients/Pinpoint/types.ts:5]: This is pretty cool, good job!

**Review by svidgen — COMMENTED** [review]
Lots of comments for you! Feel free to take most as signaling my preferences for *future* PR's. And on that note, let's try strategize future PR's to be more focused, please.

Four things I'd want touched on in this PR before approving:

1. The potential race in the response stream/body reader/cache promise.
2. Confirmation that error cases are tested to the extent they need to be.
3. A link to the encodeURIComponent RFC gapfill.
4. I had a comment somewhere in here about something that sounds like a latent bomb. Can't remember where. But, something there to settle my nerves, please!

*In general*, a few things were initially hard to follow. But, nothing that should block the PR. And, it seems like some of the harder-to-follow bits are in support of some pretty easy-to-read client/"service api" generation. So, **net good**, IMO. **Good work!**

I think most of my comments are about half-baked docstrings. If you have time, I'd love to see followup PR (at least eventually) with meaningful docstrings. It looks like you started in some areas and produced some really helpful docs and maybe just lost steam.

And my disclosure: I'm not an expert on the auth exchanges, and I didn't look to closely at the pinpoint stuff. I assume others are reviewing these in depth.

**Review by elorzafe — DISMISSED** [review]
This is awesome!!

Thanks for the hard work and putting this out.
I have some nit comments, but feel free to update later.

File: packages/core/src/AwsClients/CognitoIdentity/base.ts [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * A Cognito Identity-specific middleware that disables caching for all requests.
+ */
+const disableCacheMiddleware: Middleware<HttpRequest, HttpResponse, {}> =
+	() => (next, context) =>
+		async function disableCacheMiddleware(request) {
+			request.headers['cache-control'] = 'no-store';
+			return next(request);
+		};
```
[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply packages/core/src/AwsClients/CognitoIdentity/base.ts:42]: Because middleware is for customizations and `getSharedHeaders` are those headers defined in the service API, hence called from the serializer. Since this is a security limitation of device/client, it's added to the middleware

File: packages/core/src/clients/handlers/unauthenticated.ts [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+import { userAgentMiddleware, UserAgentOptions } from '../middleware/userAgent';
+import { composeTransferHandler } from '../internal/composeTransferHandler';
+import { fetchTransferHandler } from './fetch';
+import { HttpRequest, HttpResponse } from '../types';
+
+export const unauthenticatedHandler = composeTransferHandler<
+	[UserAgentOptions, RetryOptions<HttpResponse>],
+	HttpRequest,
+	HttpResponse
+>(fetchTransferHandler, [userAgentMiddleware, retryMiddleware]);
```
[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply packages/core/src/clients/handlers/unauthenticated.ts:14]: I think `unauthenticated` is wildly used in AWS docs, hense used here.

File: packages/core/src/clients/middleware/retry/middleware.ts [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+/**
+ * Retry middleware
+ */
+export const retryMiddleware = <TInput = Request, TOutput = Response>({
+	maxAttempts = DEFAULT_RETRY_ATTEMPTS,
+	retryDecider,
+	computeDelay,
+	abortSignal,
```
[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply packages/core/src/clients/middleware/retry/middleware.ts:50]: Yes, the composing of pure functional middleware happens for each request. It's the same for AWS SDK too.

File: packages/core/src/clients/serde/json.ts [hunk 21]
```diff
@@ -0,0 +1,53 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { ErrorParser, HttpResponse } from '../types';
+import { parseMetadata } from './responseInfo';
+
+/**
+ * Utility functions for serializing and deserializing of JSON protocol in general(including: REST-JSON, JSON-RPC, etc.)
+ * The utility functions here must be mindful of only reading the response body once for each response in any
```
[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply packages/core/src/clients/serde/json.ts:9]: Good point. This is from the older commit and should be fixed. Will remove it.

File: packages/core/src/clients/middleware/signing/utils/isClockSkewed.ts [hunk 22]
```diff
@@ -0,0 +1,26 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { getSkewCorrectedDate } from './getSkewCorrectedDate';
+
+const SKEW_WINDOW = 5 * 60 * 1000; // 5 minutes
```
[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply packages/core/src/clients/middleware/signing/utils/isClockSkewed.ts:6]: It's from the [AWS SDK](https://github.com/aws/aws-sdk-js-v3/blob/6c0f44fab30a1bb2134af47362a31332abc3666b/packages/middleware-signing/src/utils/isClockSkewed.ts#L10). Will add the reference

File: packages/core/src/clients/handlers/fetch.ts [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+ * etc.) when body is small or streaming implementation is not available(RN).
+ */
+const withPayloadCaching = <T>(payloadAccessor: () => Promise<T>) => {
+	let cached: T;
+	return async () => {
+		if (!cached) {
+			cached = await payloadAccessor();
+		}
+		return cached;
+	};
```
[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply packages/core/src/clients/handlers/fetch.ts:69]: It will not race because the cached payload is in the closure of returned function. And there will always a new closure created when `withPayloadCaching` is called.

[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply packages/core/src/clients/handlers/fetch.ts:69]: This race will only happen if the `withPayloadCaching` is called concurrently. However payload accessor is only called once per http response

File: packages/analytics/src/Providers/EventBuffer.ts (hunk 2 above)
Comment (AllanZhengYP) [reply packages/analytics/src/Providers/EventBuffer.ts:165]: Cannot make this change because of TSC prefers the current spreading.

File: packages/core/__tests__/AwsClients/CognitoIdentity/getCredentialsForIdentity-test.ts [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+			params
+		);
+		expect(response).toEqual(expectedOutput);
+		expect(fetchTransferHandler).toBeCalledWith(
+			expectedRequest,
+			expect.anything()
+		);
+	});
+
+	test('error case', async () => {
```
[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply packages/core/__tests__/AwsClients/CognitoIdentity/getCredentialsForIdentity-test.ts:78]: Yes. Because all the error from a single API follows the same error protocol. Being able to parse error code and message from one of errors ensures other errors.

File: packages/core/src/clients/middleware/retry/jitteredBackoff.ts [hunk 25]
```diff
@@ -0,0 +1,14 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import type { RetryOptions } from './middleware';
+// TODO: remove this dependency in v6
```
[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply packages/core/src/clients/middleware/retry/jitteredBackoff.ts:5]: Because it's duplicated to what we are doing with retry middleware. Other retry class should no longer relavant when we use shared http handler.

**Review by elorzafe — APPROVED** [review]
Great work

**Review by wlee221 — APPROVED** [review]
Approving for the devops portion (which got removed anyway 😛)

**Review by stocaaro — APPROVED** [review]

File: packages/core/src/clients/middleware/retry/jitteredBackoff.ts [hunk 26]
```diff
@@ -2,13 +2,15 @@
...
 
 export const jitteredBackoff: RetryOptions['computeDelay'] = attempt => {
-	const delayFunction = jitteredBackoffUtil();
+	const delayFunction = jitteredBackoffUtil(DEFAULT_MAX_DELAY_MS);
 	const delay = delayFunction(attempt);
-	return delay === false ? MAX_DELAY_MS : delay;
+	// The delayFunction returns false when the delay is greater than the max delay(5 mins).
+	// In this case, the retry middleware will delay 5 mins instead, as a ceiling of the delay.
+	return delay === false ? DEFAULT_MAX_DELAY_MS : delay;
 };
```
Comment (svidgen) [inline packages/core/src/clients/middleware/retry/jitteredBackoff.ts]: It's a bit odd having both the delay function *and* this function doing conditional things around the `DEFAULT_MAX_DELAY_MS`. But, I *think* I can see how it makes sense. And, the change addresses the larger concern I had, so I appreciate it. Thanks!

File: packages/core/src/clients/middleware/signing/utils/extendedEncodeURIComponent.ts [hunk 27]
```diff
@@ -0,0 +1,20 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+/**
+ * Wraps encodeURIComponent to encode additional characters to fully adhere to RFC 3986.
```
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/core/src/clients/middleware/signing/utils/extendedEncodeURIComponent.ts:5]: Not the kind of reference I was hoping for. But, it does demonstrates that this copies another established library that does AWS stuff. So, I'll take it! Thanks!

File: packages/core/src/clients/middleware/signing/utils/isClockSkewed.ts [hunk 28]
```diff
@@ -3,16 +3,16 @@
 
 import { getSkewCorrectedDate } from './getSkewCorrectedDate';
 
-const SKEW_WINDOW = 5 * 60 * 1000; // 5 minutes
+// 5 mins in milliseconds. Ref: https://github.com/aws/aws-sdk-js-v3/blob/6c0f44fab30a1bb2134af47362a31332abc3666b/packages/middleware-signing/src/utils/isClockSkewed.ts#L10
```
Comment (svidgen) [inline packages/core/src/clients/middleware/signing/utils/isClockSkewed.ts:6]: Not quite the explanation I thought we'd have, but it works for me!

**Review by svidgen — APPROVED** [review]
Thanks for the rev!

**Review by svidgen — APPROVED** [review]
Thanks for the rev!
